#!/bin/sh /usr/share/dpatch/dpatch-run
## 01-growisofs-pioneer.dpatch
##
## DP: Fixes to fail closing session on some PIONEER DVD-RW drives

@DPATCH@

diff -Naur dvd+rw-tools-7.0.orig/growisofs_mmc.cpp dvd+rw-tools-7.0/growisofs_mmc.cpp
--- dvd+rw-tools-7.0.orig/growisofs_mmc.cpp	2006-09-23 12:45:49.000000000 +0000
+++ dvd+rw-tools-7.0/growisofs_mmc.cpp	2006-10-06 20:48:34.000000000 +0000
@@ -1724,8 +1724,17 @@
 	cmd[1] = 0x01;		// "IMMED"
 	cmd[2] = 0x02;		// "Close session"
 	cmd[9] = 0;
-	if ((err=cmd.transport()))
-	    sperror ("CLOSE SESSION",err);
+	
+	// it seems, that pioneer is a bit crappy
+	while (err=cmd.transport()) {
+	    if (SK(err)==0x2 && ASC(err)==0x04 && ASCQ(err)==0x07) {
+		    sperror ("CLOSE SESSION (but try to continue)",err);
+		    usleep(10000);
+	    } else {
+		    sperror ("CLOSE SESSION",err);
+		    break;
+	    }
+	}
 
 	if (wait_for_unit (cmd)) break;
  
